CA2018 Analyzer Documentation (#25668)
dotnet/docs: docs/fundamentals/code-analysis/quality-rules/ca2018.md

---
title: "CA2018: `Buffer.BlockCopy` expects the number of bytes to be copied for the `count` argument"
description: "Learn about code analysis rule CA2018: `Buffer.BlockCopy` expects the number of bytes to be copied for the `count` argument"
ms.date: 08/17/2021
ms.topic: reference
f1_keywords:
  - "BufferBlockCopyLength"
  - "CA2018"
helpviewer_keywords:
  - "BufferBlockCopyLength"
  - "CA2018"
author: mahdiva
---
# CA2018: `Buffer.BlockCopy` expects the number of bytes to be copied for the `count` argument

| | Value |
|-|-|
| **Rule ID** |CA2018|
| **Category** |[Reliability](reliability-warnings.md)|
| **Fix is breaking or non-breaking** |Non-breaking|

## Cause

> This rule fires when `Array.Length` is used for the `count` argument on arrays with elements larger than 1 byte.

Reviewer: This should at least read as `elements of size larger than 1 byte`. Will leave it up to the docs folks on whether we should mention that `byte[], sbyte[] or bool[]` are the only array types where each element is of size 1 byte.

---
title: "CA2018: `Buffer.BlockCopy` expects the number of bytes to be copied for the `count` argument"
description: "Learn about code analysis rule CA2018: `Buffer.BlockCopy` expects the number of bytes to be copied for the `count` argument"
ms.date: 08/17/2021
ms.topic: reference
f1_keywords:
  - "BufferBlockCopyLength"
  - "CA2018"
helpviewer_keywords:
  - "BufferBlockCopyLength"
  - "CA2018"
author: mahdiva
---
# CA2018: `Buffer.BlockCopy` expects the number of bytes to be copied for the `count` argument

| | Value |
|-|-|
| **Rule ID** |CA2018|
| **Category** |[Reliability](reliability-warnings.md)|
| **Fix is breaking or non-breaking** |Non-breaking|

## Cause

This rule fires when `Array.Length` is used for the `count` argument on arrays with elements of size larger than 1 byte.

## Rule description

> When using `Buffer.BlockCopy`, the `count` argument expects the number of bytes to be copied. Using `Array.Length` for the `count` argument should only be used for arrays with element size of exactly 1 byte.

Reviewer 2: ```suggestion
When using `Buffer.BlockCopy`, the `count` argument specifies the number of bytes to copy. You should only use `Array.Length` for the `count` argument on arrays whose elements are exactly one byte in size. `byte`, `sbyte, and `bool` arrays have elements that are one byte in size.
```

## How to fix violations

> The `count` argument should be the number of bytes that you intend to copy.

Reviewer 2: ```suggestion
Specify the number of bytes that you intend to copy for the `count` argument.
```

### Example

> ```csharp

Reviewer 2: ```suggestion
**Violation**:

```csharp
```

using System;
class Program
{
    static void Main()
    {
        int[] src = new int[] {1, 2, 3, 4};
        int[] dst = new int[] {0, 0, 0, 0};

        Buffer.BlockCopy(src, 0, dst, 0, src.Length); 
    }
}
```

> Fix:

Reviewer 2: ```suggestion
**Fix**:
```

> When using arrays with elements of size larger than 1 byte, one possible solution is to multiply the length of the array with the array element size to get the number of bytes.

Reviewer 2: ```suggestion
If your array's elements are larger than one byte in size, you can multiply the length of the array by the element size to get the number of bytes.
```
